vernac/lemmas.ml

```diff
@@ -196,121 +189,92 @@ let retrieve_first_recthm uctx = function
+    let _, de = DeclareDef.prepare_definition ~allow_evars:false ~opaque ~poly sigma udecl ~types:(Some typ) ~body in
+    let hook_data = Option.map (fun hook -> hook, uctx, []) hook in
+    let ubind = UnivNames.empty_binders in (* XXX fixme ubind *)
+    DeclareDef.declare_definition ~name ~scope ~kind ?hook_data ubind de impargs
+
+(* This declares implicits and calls the hooks for other_thms *)
+let process_recthms ?fix_exn ?hook uctx ~udecl ~poly ~scope dref other_thms =
+  if List.is_empty other_thms then ()
+  else
+    let body, opaque = retrieve_first_recthm uctx dref in
+    let _ = List.map_i (save_remaining_recthms ?fix_exn ~poly ~scope ~udecl ~hook uctx body opaque) 1 other_thms in
+    ()
 
 (************************************************************************)
 (* Admitting a lemma-like constant                                      *)
 (************************************************************************)
 
 (* Admitted *)
-let warn_let_as_axiom =
-  CWarnings.create ~name:"let-as-axiom" ~category:"vernacular"
-                   (fun id -> strbrk "Let definition" ++ spc () ++ Id.print id ++
-                                spc () ++ strbrk "declared as an axiom.")
-
 let get_keep_admitted_vars =
   Goptions.declare_bool_option_and_ref
     ~depr:false
     ~key:["Keep"; "Admitted"; "Variables"]
     ~value:true
 
-let finish_admitted env sigma ~name ~poly ~scope pe ctx hook ~udecl impargs other_thms =
-  let open DeclareDef in
-  let local = match scope with
-  | Global local -> local
-  | Discharge -> warn_let_as_axiom name; Declare.ImportNeedQualified
-  in
-  let kn = Declare.declare_constant ~name ~local ~kind:Decls.(IsAssumption Conjectural) (Declare.ParameterEntry pe) in
+let compute_proof_using_for_admitted proof typ pproofs =
+  if not (get_keep_admitted_vars ()) then None
+  else match Proof_global.get_used_variables proof, pproofs with
+    | Some _ as x, _ -> x
+    | None, pproof :: _ ->
+      let env = Global.env () in
+      let ids_typ = Environ.global_vars_set env typ in
+      (* [pproof] is evar-normalized by [partial_proof]. We don't
+         count variables appearing only in the type of evars. *)
+      let ids_def = Environ.global_vars_set env (EConstr.Unsafe.to_constr pproof) in
+      Some (Environ.really_needed env (Id.Set.union ids_typ ids_def))
+    | _ -> None
+
+let finish_admitted ~name ~poly ~scope pe uctx hook ~udecl impargs other_thms =
+  let dref = DeclareDef.declare_assumption ~name ~scope ~hook ~uctx ~impargs pe in
+  (* Should both of those be done in declare_axiom? *)
   let () = Declare.assumption_message name in
-  DeclareUniv.declare_univ_binders (GlobRef.ConstRef kn) (UState.universe_binders ctx);
-  (* This takes care of the implicits and hook for the current constant*)
-  process_recthms ?fix_exn:None ?hook env sigma ctx ~udecl ~poly ~scope:(Global local) (GlobRef.ConstRef kn) impargs other_thms
+  DeclareUniv.declare_univ_binders dref (UState.universe_binders uctx);
```

Comment:
Declaredef should handle that.

Reply:
I added a comment; indeed the problem again is that old code was not doing this for the rest of the constants, so if we move this `DeclareDef` now we will different semantics for mutual assumptions [how is that even handled tho]

I'm gonna push a commit with you suggestion and see what happens.

Reply:
Not sure the case admit + recthms is possible tho?